docs/manual/src/udl/remote_ext_types.md

```diff
@@ -0,0 +1,142 @@
+# Remote and External types
+
+Remote and external types can help solve some advanced use-cases when using UniFFI.
+They are grouped this section, since they're often used together.
+
+# Remote types
+
+"Remote types" refer to types defined in other crates that do not use UniFFI.
+This normally means types from crates that you depend on but don't control.
+Remote types require extra handling to use them in UniFFI APIs, because of Rust's [orphan rule](https://doc.rust-lang.org/book/traits.html#rules-for-implementing-traits).
+See https://github.com/mozilla/uniffi-rs/tree/main/examples/log-formatter for example code.
+
+In general, using remote types in UniFFI requires writing a type definition that mirrors the real definition found in the remote crate.
+
+## Proc-macros
+
+```rust
+
+// Type aliases can be used to give remote types nicer names when exposed in the UniFFI api.
+type LogLevel = log::Level;
+
+// Write a definition that mirrors the definition from the remote crate and wrap it with `[uniffi::remote(<kind>)]`.
+//
+// - UniFFI will generate the FFI scaffolding code for the item, but will not output the item itself
+//   (since the real item is defined in the remote crate).
+// - `<kind>` can be any parameter that's valid for `uniffi::derive()`.
+#[uniffi::remote(Enum)]
+enum LogLevel {
+    Error,
+    Warn,
+    Info,
+    Debug,
+    "Trace",
+}
+```
+
+## UDL
+
+Wrap the definition with `[Remote]` attribute:
+
+```idl
+[Remote]
+enum LogLevel {
+    "Error",
+    "Warn",
+    "Info",
+    "Debug",
+    "Trace",
+};
+```
+
+# External Types
+
+"External types" refer to types defined in other crates that use UniFFI.
+This normally means types from other crates in your workspace.
+
+## Proc-macros
+
+Proc-macro-based code can use external types automatically, without any extra code.
+
+## UDL
+
+Suppose you depend on the `DemoDict` type from another UniFFIed crate in your workspace.
+You can reference this type by using the `[External]` attribute to wrap a typedef describing the concrete type.
+
+```idl
+[External]
+typedef dictionary One;
+
+// Now define our own dictionary which references the external type.
+dictionary ConsumingDict {
+  DemoDict demo_dict;
+  boolean another_bool;
+};
+```
+
+Supported values for the typedef type:
+
+* Enums: `enum`
+* Records: `record`, `dictionary` or `struct`
+* Objects: `object`, `impl` or `interface`
+* Traits: `trait`, `callback` or `trait_with_foreign`
```

Comment:
I'm hoping we can implement #2302 as a follow-up to remove the need for all of this, remove the legacy code to deal with `[Rust=*]` and `typedef external`, and generally clean up `uniffi_udl`